book/src/developers/designs/oauth2_refresh_tokens.md

```diff
@@ -0,0 +1,161 @@
+          │   │                           │         │       │      ▲                        │         │
+          └───┼───────────────────────────┼─────────┘       └──────┼────────────────────────┼─────────┘
+         ┌────┘             ▲        ┌────┘                        │                   ┌────┘          
+         │                  │        │                             │                   │               
+         │                  │        │                             │                   │               
+    1. Issued               │   5. Issued                          │              8. Issued            
+         │                  │        │                             │                   │               
+         │                  │        │                             │                   │               
+         │                  │        │                             │                   │               
+         ▼                  │        ▼                             │                   ▼               
+     ┌───────┐              │    ┌───────┐                         │               ┌───────┐           
+     │       │              │    │       │                         │               │       │           
+     │Access │              │    │Access │                         │               │Access │           
+     │   +   │              │    │   +   │                         │               │   +   │           
+     │Refresh│──3. Refresh──┘    │Refresh│                         │               │Refresh│           
+     │ IAT 1 │                   │ IAT 2 │─────6. Refresh──────────┘               │ IAT 1 │           
+     │       │                   │       │                                         │       │           
+     └───────┘                   └───────┘                                         └───────┘
+
+In this example, we can see that the replication of the session with NIB 1 happens to the second
+Kanidm server, but the replication of session with NIB 2 has not occurred yet. If the token that was
+later issued with IAT 2 was presented to the second server it would still be valid and able to
+refresh since IAT 2 is greater or equal to NIB 1. This would also prompt the session to advance to
+NIB 3 such that when replication begun again, the session with NIB 3 would take precedence over the
+former NIB 2 session.
+
+While this allows a short window where a former access token could be used on the second replica,
+this infrastructure being behind load balancers and outside of an attackers influence significantly
+hinders the ability to attack this for very little gain.
+
+## Attack Detection
+
+[draft oauth security topics section 4.14.2](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#section-4.14.2)
+specifically calls out that when refresh token re-use is detected then all tokens of the session
+should be canceled to cause a new authorisation code flow to be initiated.
```

Comment:
This is definitely good to include, and I'd specifically call out an audit/alarm event (once it's implemented).

Reply:
Yep agreed. I'm going to add some tests for this too.